resource/param/preparer.go

```diff
@@ -15,40 +15,186 @@
+// fragment as input. The validation logic then wraps the expression around a
+// different template fragment, then evals it to determine the outcome
+
 // Preparer for params
 //
 // Param controls the flow of values through `module` calls. You can use the
 // `{{param "name"}}` template call anywhere you need the value of a param
 // inside the current module.
 type Preparer struct {
-	// Default is an optional field that provides a default value if none is
-	// provided to this parameter. If this field is not set, this param will be
-	// treated as required.
+	// Default is an optional field that provides a default value if none
+	// is provided to this parameter. If this field is not set, this param
+	// will be treated as required. If this is provided, then Type is
+	// inferred
 	Default *string `hcl:"default"`
+
+	// Type is an optional field that is used for type checking. As of
+	// right now, it can either be ParamTypeString or ParamTypeInt. It
+	// is used with typeCastValue.
+	Type ParamType `hcl:"type"`
+
+	Must []string `hcl:"must"`
 }
 
 // Prepare a new task
 func (p *Preparer) Prepare(render resource.Renderer) (resource.Task, error) {
-	if val, present := render.Value(); present {
-		return &Param{Value: val}, nil
+	value, present := render.Value()
+
+	if !present {
+		if p.Default == nil {
+			return nil, errors.Errorf("param is required")
+		}
+		def, err := render.Render("default", *p.Default)
+		if err != nil {
+			return nil, err //errors.Wrap(err, "rendering default failed")
+		}
+		value = def
 	}
 
-	if p.Default == nil {
-		return nil, errors.New("param is required")
+	typedValue, err := typeCastValue(value, p.Type)
+	if err != nil {
+		return nil, errors.Wrap(err, "typecasting value failed")
 	}
 
-	def, err := render.Render("default", *p.Default)
+	err = ValidateType(typedValue, p.Predicates())
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "param validation failed")
+	}
+
+	return &Param{Value: value}, nil
+}
+
+func (p *Preparer) Predicates() map[string]string {
+
+	predicates := make(map[string]string)
+
+	for _, rule := range p.Must {
+		count := len(predicates)
+		name := fmt.Sprintf("pred#%d", count)
+		predicate := fmt.Sprintf("{{if %s }}%d{{else}}%d{{end}}", rule, ValidatePass, ValidateFail)
+		predicates[name] = predicate
 	}
 
-	return &Param{Value: def}, nil
+	return predicates
+}
+
+func typeCastValue(paramValue string, paramType ParamType) (interface{}, error) {
+	switch paramType {
+	case ParamTypeInt:
+		value, err := strconv.Atoi(paramValue)
+		if err != nil {
+			return value, errors.Wrapf(err, `paramType is "int", but converting "%s" failed`, paramValue)
+		}
+		return value, nil
+
+	// this case is a nop, since string is the default type for parameters
+	case ParamTypeString:
+		return paramValue, nil
+	case ParamTypeInferred:
+		// try to infer the type from the value
+		// the recursion is make the code DRYer, but is this the best way?
```

Comment:
I could see this getting unwieldy if we want to add more types for validation. I wonder if we just require that the `type` is specified when using validation in order to keep things simple. Or, perhaps better, maybe it's possible to build on the work in #321 to infer the type here.